go-controller/pkg/ovn/namespace.go

```diff
@@ -376,18 +390,23 @@ func (oc *Controller) getNamespaceLocked(ns string) *namespaceInfo {
 	return nsInfo
 }
 
-// createNamespaceLocked locks namespacesMutex, creates an entry for ns, and returns it
+// ensureNamespaceLocked locks namespacesMutex, gets/creates an entry for ns, and returns it
 // with its mutex locked.
-func (oc *Controller) createNamespaceLocked(ns string) *namespaceInfo {
+func (oc *Controller) ensureNamespaceLocked(ns string) *namespaceInfo {
 	oc.namespacesMutex.Lock()
```

Comment:
this probably has a lock inversion problem. With multiple go-routines running for the queued informer, we will have the oc.namespacesMutex lock being waited on when another go-routine can wait on the nsInfo.lock() for an nsInfo which a third go-routine has grabbed a lock on using `getNamespaceInfoLocked`.

Reply:
Why don't we just return nsInfo by value?

Reply:
All we need to make sure is the operations that modify the content of nsInfo for each namespace are locked.

Reply:
I think I fixed the locking problem. When the nsInfo existed, I was accidentally holding the namespacesMutex and trying to nsInfo Lock at the same time. Re-worked it to follow the getNamespacesLocked logic where we try to obtain nsInfo Lock after unlocking namespacesMutex, then double checking to make sure it still exists. It's not pretty but it seems to have remedied the issue.

Reply:
> When the nsInfo existed, I was accidentally holding the namespacesMutex and trying to nsInfo Lock at the same time

why would it be safe here? like even in the case `nsinfo` didn't exist yet, you'd still now ultimately be holding the mutex lock AND the nsinfo lock at line 412 right? The only difference being that you created + locked `nsinfo` rather than just getting and locking `nsinfo`

Reply:
in the case where nsinfo didnt exist, we can safely hold both locks because no one else could possibly be trying to lock that nsinfo (it doesnt exist yet).

 In the case where it exists already, someone else could be holding its lock and have released the namespace mutex, but trying to also obtain namespace mutex lock (like in getNamespaceLocked). If this happens and ensureNamespaceLocked is running, you could have deadlock. Take this example with thread A and B, and a single nsInfo already exists.

1. thread A calls getNamespaceLocked, in that function, it locks namespaceMutex, gets nsinfo, unlocks namespaceMutex, locks nsinfo
2. thread B calls ensureNamespaceLocked, in that function, it locks namespaceMutex, tries to lock nsinfo
3. thread A is finishing the getNamespaceLocked function, so it is doing:
```
	// Check that the namespace wasn't deleted while we were waiting for the lock
	oc.namespacesMutex.Lock()
	defer oc.namespacesMutex.Unlock()
	if nsInfo != oc.namespaces[ns] {
		nsInfo.Unlock()
		return nil
	}
```
thread A is unable to lock namespacesMutex cause it is held by thread B. Thread B is unable to get nsInfo lock because it is held by thread A. Now both threads are deadlocked.